File: addons/godot_gameinput_csharp/GameInputReading.cs
using System;
using System.Collections.Generic;
using Godot;
using GodotGameInput.Internal;

namespace GodotGameInput;

/// <summary>
/// Immutable snapshot of one device's input, mirroring the native
/// <c>GameInputReading</c>. Covers every input kind the device supports
/// (gamepad, keyboard, mouse, sensors, arcade stick, flight stick, racing wheel
/// and raw controller) and holds the state it replaced, so edge queries return
/// correct transitions. Methods for a kind the reading does not contain return
/// false, 0 or an empty value; check <see cref="InputKinds"/>.
/// </summary>
public sealed class GameInputReading : GameInputObject
{
    internal GameInputReading(GodotObject o) : base(o)
    {
    }

    public static GameInputReading From(GodotObject o) => o == null ? null : new GameInputReading(o);

    // --- Gamepad ---

    /// <summary>True when every bit of <paramref name="button"/> is held (a chord for several flags).</summary>
    public bool IsButtonDown(GameInputDevice.Button button) => Call("is_button_down", (int)button).AsBool();

    public bool IsButtonDown(int buttonMask) => Call("is_button_down", buttonMask).AsBool();

    public bool WasButtonPressed(GameInputDevice.Button button) =>
        Call("was_button_pressed", (int)button).AsBool();

    public bool WasButtonReleased(GameInputDevice.Button button) =>
        Call("was_button_released", (int)button).AsBool();

    /// <summary>Axis value; thumbstick Y is down-positive like Godot's joypad axes.</summary>
    public float GetAxis(GameInputDevice.Axis axis) => (float)Call("get_axis", (int)axis).AsDouble();

    public int GetButtonsMask() => Call("get_buttons_mask").AsInt32();

    // --- Metadata ---

    /// <summary>Microseconds on the GameInput clock; compare with <see cref="GameInput.CurrentTimestamp"/>.</summary>
    public long GetTimestamp() => Call("get_timestamp").AsInt64();

    public long DeviceId => Call("get_device_id").AsInt64();

    /// <summary>Timestamp of the part of the reading holding <paramref name="kind"/> (newest for a mask).</summary>
    public long GetKindTimestamp(GameInput.DeviceKind kind) => Call("get_kind_timestamp", (int)kind).AsInt64();

    /// <summary>The <see cref="GameInput.DeviceKind"/> flags of the input this reading contains.</summary>
    public GameInput.DeviceKind InputKinds => (GameInput.DeviceKind)Call("get_input_kinds").AsInt32();

    /// <summary>False for the first reading after a device connects: release edges need a previous state.</summary>
    public bool HasPrevious => Call("has_previous").AsBool();

    /// <summary>True when event-driven readings of this device were dropped just before this one.</summary>
    public bool HasGapBefore => Call("has_gap_before").AsBool();

    /// <summary>True when the device reported more keys, axes, buttons or switches than a reading holds.</summary>
    public bool IsTruncated => Call("is_truncated").AsBool();

    // --- Unified sources (any device kind) ---

    public bool IsSourceDown(GameInputDevice.Source source) => Call("is_source_down", (int)source).AsBool();

    public bool WasSourcePressed(GameInputDevice.Source source) =>
        Call("was_source_pressed", (int)source).AsBool();

    public bool WasSourceReleased(GameInputDevice.Source source) =>
        Call("was_source_released", (int)source).AsBool();

    /// <summary>Axis value for axis sources; 1 or 0 for button sources.</summary>
    public float GetSourceValue(GameInputDevice.Source source) =>
        (float)Call("get_source_value", (int)source).AsDouble();

    // --- Keyboard ---

    public int KeyCount => Call("get_key_count").AsInt32();

    /// <summary>Physical (position-based) <see cref="Key"/> values of the held keys, like <c>InputEventKey.PhysicalKeycode</c>.</summary>
    public Key[] GetPressedPhysicalKeys() =>
        Array.ConvertAll(Call("get_pressed_physical_keys").AsInt64Array(), k => (Key)k);

    public bool IsPhysicalKeyDown(Key physicalKey) => Call("is_physical_key_down", (long)physicalKey).AsBool();

    public bool WasPhysicalKeyPressed(Key physicalKey) =>
        Call("was_physical_key_pressed", (long)physicalKey).AsBool();

    public bool WasPhysicalKeyReleased(Key physicalKey) =>
        Call("was_physical_key_released", (long)physicalKey).AsBool();

    /// <summary>
    /// One dictionary per held key: <c>scan_code</c>, <c>virtual_key</c>,
    /// <c>code_point</c>, <c>is_dead_key</c>, <c>physical_keycode</c>,
    /// <c>keycode</c> and <c>location</c>.
    /// </summary>
    public IReadOnlyList<Godot.Collections.Dictionary> GetKeyStates()
    {
        var result = new List<Godot.Collections.Dictionary>();
        foreach (Variant state in Call("get_key_states").AsGodotArray())
        {
            result.Add(state.AsGodotDictionary());
        }

        return result;
    }

    // --- Mouse ---

    public GameInputDevice.MouseButton MouseButtons =>
        (GameInputDevice.MouseButton)Call("get_mouse_buttons").AsInt32();

    public bool IsMouseButtonDown(GameInputDevice.MouseButton button) =>
        Call("is_mouse_button_down", (int)button).AsBool();

    public bool WasMouseButtonPressed(GameInputDevice.MouseButton button) =>
        Call("was_mouse_button_pressed", (int)button).AsBool();

    public bool WasMouseButtonReleased(GameInputDevice.MouseButton button) =>
        Call("was_mouse_button_released", (int)button).AsBool();

    /// <summary>Mouse motion since the previous state, in device units (not screen pixels).</summary>
    public Vector2 MouseDelta => Call("get_mouse_delta").AsVector2();

    /// <summary>Horizontal (x) and vertical (y) wheel motion since the previous state.</summary>
    public Vector2 MouseWheelDelta => Call("get_mouse_wheel_delta").AsVector2();

    public bool HasMouseAbsolutePosition => Call("has_mouse_absolute_position").AsBool();

    public Vector2 MouseAbsolutePosition => Call("get_mouse_absolute_position").AsVector2();

    /// <summary>Raw accumulated mouse state (64-bit wrapping counters); prefer <see cref="MouseDelta"/>.</summary>
    public Godot.Collections.Dictionary GetMouseState() => Call("get_mouse_state").AsGodotDictionary();

    // --- Motion sensors ---

    public GameInputDevice.SensorKind SensorKinds => (GameInputDevice.SensorKind)Call("get_sensor_kinds").AsInt32();

    /// <summary>Acceleration in metres per second squared.</summary>
    public Vector3 Accelerometer => Call("get_accelerometer").AsVector3();

    /// <summary>Angular velocity in radians per second.</summary>
    public Vector3 Gyroscope => Call("get_gyroscope").AsVector3();

    public float HeadingDegrees => (float)Call("get_heading_degrees").AsDouble();

    public GameInputDevice.SensorAccuracy HeadingAccuracy =>
        (GameInputDevice.SensorAccuracy)Call("get_heading_accuracy").AsInt32();

    public Quaternion Orientation => Call("get_orientation").AsQuaternion();

    // --- Arcade stick, flight stick, racing wheel ---

    public GameInputDevice.ArcadeStickButton ArcadeStickButtons =>
        (GameInputDevice.ArcadeStickButton)Call("get_arcade_stick_buttons").AsInt32();

    public GameInputDevice.FlightStickButton FlightStickButtons =>
        (GameInputDevice.FlightStickButton)Call("get_flight_stick_buttons").AsInt32();

    public GameInputDevice.SwitchPosition FlightStickHat =>
        (GameInputDevice.SwitchPosition)Call("get_flight_stick_hat").AsInt32();

    public Vector2 FlightStickHatVector => Call("get_flight_stick_hat_vector").AsVector2();

    public GameInputDevice.RacingWheelButton RacingWheelButtons =>
        (GameInputDevice.RacingWheelButton)Call("get_racing_wheel_buttons").AsInt32();

    public int RacingWheelGear => Call("get_racing_wheel_gear").AsInt32();

    // --- Raw controller ---

    /// <summary>Every raw controller axis, normalized by GameInput to [0, 1]; the layout is device-specific.</summary>
    public float[] GetControllerAxes() => Call("get_controller_axes").AsFloat32Array();

    public bool[] GetControllerButtons()
    {
        Godot.Collections.Array buttons = Call("get_controller_buttons").AsGodotArray();
        var result = new bool[buttons.Count];
        for (int i = 0; i < result.Length; i++)
        {
            result[i] = buttons[i].AsBool();
        }

        return result;
    }

    public GameInputDevice.SwitchPosition[] GetControllerSwitches() =>
        Array.ConvertAll(Call("get_controller_switches").AsInt32Array(), s => (GameInputDevice.SwitchPosition)s);

    public float GetControllerAxis(int index) => (float)Call("get_controller_axis", index).AsDouble();

    public bool IsControllerButtonDown(int index) => Call("is_controller_button_down", index).AsBool();

    public bool WasControllerButtonPressed(int index) => Call("was_controller_button_pressed", index).AsBool();

    public bool WasControllerButtonReleased(int index) => Call("was_controller_button_released", index).AsBool();

    public GameInputDevice.SwitchPosition GetControllerSwitch(int index) =>
        (GameInputDevice.SwitchPosition)Call("get_controller_switch", index).AsInt32();
}

File: addons/godot_gameinput/src/gameinput_reading.cpp
#include "gameinput_reading.h"

#include "gameinput_device.h"
#include "gameinput_keymap.h"
#include "gameinput_singleton.h"

#include <cmath>

namespace godot {

namespace gi = gameinput_internal;

namespace {

constexpr uint32_t kMouseButtonMask = 0x7Fu;
constexpr uint32_t kSpecialtyButtonMask = 0x3FFFu;
constexpr float kAxisSourceThreshold = 0.5f;
constexpr float kStandardGravity = 9.80665f;

int source_bit_index(int source, int base, int count) {
    if (source < base || source >= base + count) return -1;
    return source - base;
}

} // namespace

void GameInputReading::_bind_methods() {
    ClassDB::bind_method(D_METHOD("is_button_down", "button"), &GameInputReading::is_button_down);
    ClassDB::bind_method(D_METHOD("was_button_pressed", "button"),
                         &GameInputReading::was_button_pressed);
    ClassDB::bind_method(D_METHOD("was_button_released", "button"),
                         &GameInputReading::was_button_released);
    ClassDB::bind_method(D_METHOD("get_axis", "axis"), &GameInputReading::get_axis);
    ClassDB::bind_method(D_METHOD("get_buttons_mask"), &GameInputReading::get_buttons_mask);
    ClassDB::bind_method(D_METHOD("get_timestamp"), &GameInputReading::get_timestamp);
    ClassDB::bind_method(D_METHOD("get_kind_timestamp", "kind"),
                         &GameInputReading::get_kind_timestamp);
    ClassDB::bind_method(D_METHOD("get_device_id"), &GameInputReading::get_device_id);
    ClassDB::bind_method(D_METHOD("get_input_kinds"), &GameInputReading::get_input_kinds);
    ClassDB::bind_method(D_METHOD("has_previous"), &GameInputReading::has_previous);
    ClassDB::bind_method(D_METHOD("has_gap_before"), &GameInputReading::has_gap_before);
    ClassDB::bind_method(D_METHOD("is_truncated"), &GameInputReading::is_truncated);

    ClassDB::bind_method(D_METHOD("is_source_down", "source"), &GameInputReading::is_source_down);
    ClassDB::bind_method(D_METHOD("was_source_pressed", "source"),
                         &GameInputReading::was_source_pressed);
    ClassDB::bind_method(D_METHOD("was_source_released", "source"),
                         &GameInputReading::was_source_released);
    ClassDB::bind_method(D_METHOD("get_source_value", "source"),
                         &GameInputReading::get_source_value);

    ClassDB::bind_method(D_METHOD("get_key_count"), &GameInputReading::get_key_count);
    ClassDB::bind_method(D_METHOD("get_pressed_physical_keys"),
                         &GameInputReading::get_pressed_physical_keys);
    ClassDB::bind_method(D_METHOD("is_physical_key_down", "physical_key"),
                         &GameInputReading::is_physical_key_down);
    ClassDB::bind_method(D_METHOD("was_physical_key_pressed", "physical_key"),
                         &GameInputReading::was_physical_key_pressed);
    ClassDB::bind_method(D_METHOD("was_physical_key_released", "physical_key"),
                         &GameInputReading::was_physical_key_released);
    ClassDB::bind_method(D_METHOD("get_key_states"), &GameInputReading::get_key_states);

    ClassDB::bind_method(D_METHOD("get_mouse_buttons"), &GameInputReading::get_mouse_buttons);
    ClassDB::bind_method(D_METHOD("is_mouse_button_down", "button"),
                         &GameInputReading::is_mouse_button_down);
    ClassDB::bind_method(D_METHOD("was_mouse_button_pressed", "button"),
                         &GameInputReading::was_mouse_button_pressed);
    ClassDB::bind_method(D_METHOD("was_mouse_button_released", "button"),
                         &GameInputReading::was_mouse_button_released);
    ClassDB::bind_method(D_METHOD("get_mouse_delta"), &GameInputReading::get_mouse_delta);
    ClassDB::bind_method(D_METHOD("get_mouse_wheel_delta"),
                         &GameInputReading::get_mouse_wheel_delta);
    ClassDB::bind_method(D_METHOD("get_mouse_absolute_position"),
                         &GameInputReading::get_mouse_absolute_position);
    ClassDB::bind_method(D_METHOD("has_mouse_absolute_position"),
                         &GameInputReading::has_mouse_absolute_position);
    ClassDB::bind_method(D_METHOD("get_mouse_state"), &GameInputReading::get_mouse_state);

    ClassDB::bind_method(D_METHOD("get_sensor_kinds"), &GameInputReading::get_sensor_kinds);
    ClassDB::bind_method(D_METHOD("get_accelerometer"), &GameInputReading::get_accelerometer);
    ClassDB::bind_method(D_METHOD("get_gyroscope"), &GameInputReading::get_gyroscope);
    ClassDB::bind_method(D_METHOD("get_heading_degrees"), &GameInputReading::get_heading_degrees);
    ClassDB::bind_method(D_METHOD("get_heading_accuracy"),
                         &GameInputReading::get_heading_accuracy);
    ClassDB::bind_method(D_METHOD("get_orientation"), &GameInputReading::get_orientation);

    ClassDB::bind_method(D_METHOD("get_arcade_stick_buttons"),
                         &GameInputReading::get_arcade_stick_buttons);
    ClassDB::bind_method(D_METHOD("get_flight_stick_buttons"),
                         &GameInputReading::get_flight_stick_buttons);
    ClassDB::bind_method(D_METHOD("get_flight_stick_hat"), &GameInputReading::get_flight_stick_hat);
    ClassDB::bind_method(D_METHOD("get_flight_stick_hat_vector"),
                         &GameInputReading::get_flight_stick_hat_vector);
    ClassDB::bind_method(D_METHOD("get_racing_wheel_buttons"),
                         &GameInputReading::get_racing_wheel_buttons);
    ClassDB::bind_method(D_METHOD("get_racing_wheel_gear"),
                         &GameInputReading::get_racing_wheel_gear);

    ClassDB::bind_method(D_METHOD("get_controller_axes"), &GameInputReading::get_controller_axes);
    ClassDB::bind_method(D_METHOD("get_controller_buttons"),
                         &GameInputReading::get_controller_buttons);
    ClassDB::bind_method(D_METHOD("get_controller_switches"),
                         &GameInputReading::get_controller_switches);
    ClassDB::bind_method(D_METHOD("get_controller_axis", "index"),
                         &GameInputReading::get_controller_axis);
    ClassDB::bind_method(D_METHOD("is_controller_button_down", "index"),
                         &GameInputReading::is_controller_button_down);
    ClassDB::bind_method(D_METHOD("was_controller_button_pressed", "index"),
                         &GameInputReading::was_controller_button_pressed);
    ClassDB::bind_method(D_METHOD("was_controller_button_released", "index"),
                         &GameInputReading::was_controller_button_released);
    ClassDB::bind_method(D_METHOD("get_controller_switch", "index"),
                         &GameInputReading::get_controller_switch);
}

void GameInputReading::_set_snapshot(int64_t device_id, const gi::Snapshot &cur,
                                     const gi::Snapshot &prev, uint32_t prev_kinds,
                                     bool gap_before, uint32_t sensor_kinds) {
    m_device_id = device_id;
    m_cur = cur;
    m_prev = prev;
    m_prev_kinds = prev_kinds & prev.kinds;
    m_gap_before = gap_before;
    m_sensor_kinds = sensor_kinds;
}

// --- General ---------------------------------------------------------------

bool GameInputReading::is_button_down(int button) const {
    return gi::chord_down(m_cur.gamepad.buttons, (uint32_t)button,
                          GameInputDevice::kGamepadButtonMask);
}

bool GameInputReading::was_button_pressed(int button) const {
    return gi::chord_pressed(m_cur.gamepad.buttons, m_prev.gamepad.buttons,
                             _has_prev_for(gi::SNAP_GAMEPAD), (uint32_t)button,
                             GameInputDevice::kGamepadButtonMask);
}

bool GameInputReading::was_button_released(int button) const {
    return gi::chord_released(m_cur.gamepad.buttons, m_prev.gamepad.buttons,
                              _has_prev_for(gi::SNAP_GAMEPAD), (uint32_t)button,
                              GameInputDevice::kGamepadButtonMask);
}

float GameInputReading::_axis_in(const gi::Snapshot &s, int axis) const {
    using GD = GameInputDevice;
    switch (axis) {
        case GD::AXIS_LEFT_X:          return s.gamepad.left_x;
        // Godot Y axis convention: down is positive; GameInput thumbstick Y up is positive.
        // 0 - y rather than -y so a resting stick reads 0.0, not -0.0 ("-0.00" when printed).
        case GD::AXIS_LEFT_Y:          return 0.0f - s.gamepad.left_y;
        case GD::AXIS_RIGHT_X:         return s.gamepad.right_x;
        case GD::AXIS_RIGHT_Y:         return 0.0f - s.gamepad.right_y;
        case GD::AXIS_LEFT_TRIGGER:    return s.gamepad.left_trigger;
        case GD::AXIS_RIGHT_TRIGGER:   return s.gamepad.right_trigger;
        case GD::AXIS_WHEEL:           return s.racing_wheel.wheel;
        case GD::AXIS_THROTTLE:        return s.racing_wheel.throttle;
        case GD::AXIS_BRAKE:           return s.racing_wheel.brake;
        case GD::AXIS_CLUTCH:          return s.racing_wheel.clutch;
        case GD::AXIS_HANDBRAKE:       return s.racing_wheel.handbrake;
        case GD::AXIS_FLIGHT_ROLL:     return s.flight_stick.roll;
        case GD::AXIS_FLIGHT_PITCH:    return s.flight_stick.pitch;
        case GD::AXIS_FLIGHT_YAW:      return s.flight_stick.yaw;
        case GD::AXIS_FLIGHT_THROTTLE: return s.flight_stick.throttle;
        default:                       return 0.0f;
    }
}

float GameInputReading::get_axis(int axis) const {
    return _axis_in(m_cur, axis);
}

int GameInputReading::get_buttons_mask() const {
    return (int)m_cur.gamepad.buttons;
}

int64_t GameInputReading::get_timestamp() const {
    return (int64_t)gi::newest_timestamp(m_cur);
}

int64_t GameInputReading::get_kind_timestamp(int kind) const {
    return (int64_t)gi::kinds_timestamp(m_cur, gi::snapshot_kinds_from_public((uint32_t)kind));
}

int GameInputReading::get_input_kinds() const {
    return (int)gi::public_kinds_from_snapshot(m_cur.kinds);
}

bool GameInputReading::has_previous() const {
    return (m_prev_kinds & m_cur.kinds) != 0;
}

bool GameInputReading::is_truncated() const {
    return gi::is_truncated(m_cur);
}

// --- Sources -----------------------------------------------------------------

uint32_t GameInputReading::_source_snap_kind(int source) const {
    if (source >= GameInputDevice::SRC_BTN_MENU && source <= GameInputDevice::SRC_BTN_PADDLE_RIGHT_2) {
        return gi::SNAP_GAMEPAD;
    }
    if (source >= GameInputDevice::SRC_AXIS_LEFT_X && source <= GameInputDevice::SRC_AXIS_RIGHT_TRIGGER) {
        return gi::SNAP_GAMEPAD;
    }
    if (source >= GameInputDevice::SRC_AXIS_WHEEL && source <= GameInputDevice::SRC_AXIS_HANDBRAKE) {
        return gi::SNAP_RACING_WHEEL;
    }
    if (source >= GameInputDevice::SRC_AXIS_FLIGHT_ROLL &&
            source <= GameInputDevice::SRC_AXIS_FLIGHT_THROTTLE) {
        return gi::SNAP_FLIGHT_STICK;
    }
    if (source >= GameInputDevice::SRC_ARCADE_MENU && source <= GameInputDevice::SRC_ARCADE_SPECIAL_2) {
        return gi::SNAP_ARCADE_STICK;
    }
    if (source >= GameInputDevice::SRC_FLIGHT_MENU &&
            source <= GameInputDevice::SRC_FLIGHT_RIGHT_SHOULDER) {
        return gi::SNAP_FLIGHT_STICK;
    }
    if (source >= GameInputDevice::SRC_WHEEL_MENU && source <= GameInputDevice::SRC_WHEEL_RIGHT_THUMB) {
        return gi::SNAP_RACING_WHEEL;
    }
    return 0;
}

bool GameInputReading::_source_down_in(const gi::Snapshot &s, int source) const {
    int bit = source_bit_index(source, GameInputDevice::SRC_BTN_MENU, 30);
    if (bit >= 0) return ((s.gamepad.buttons >> bit) & 1u) != 0;
    bit = source_bit_index(source, GameInputDevice::SRC_ARCADE_MENU, 14);
    if (bit >= 0) return ((s.arcade_stick.buttons >> bit) & 1u) != 0;
    bit = source_bit_index(source, GameInputDevice::SRC_FLIGHT_MENU, 14);
    if (bit >= 0) return ((s.flight_stick.buttons >> bit) & 1u) != 0;
    bit = source_bit_index(source, GameInputDevice::SRC_WHEEL_MENU, 14);
    if (bit >= 0) return ((s.racing_wheel.buttons >> bit) & 1u) != 0;
    int axis = source_bit_index(source, GameInputDevice::SRC_AXIS_LEFT_X, 15);
    if (axis >= 0) return std::fabs(_axis_in(s, axis)) >= kAxisSourceThreshold;
    return false;
}

bool GameInputReading::is_source_down(int source) const {
    if (!(m_cur.kinds & _source_snap_kind(source))) return false;
    return _source_down_in(m_cur, source);
}

bool GameInputReading::was_source_pressed(int source) const {
    uint32_t kind = _source_snap_kind(source);
    if (!(m_cur.kinds & kind) || !_source_down_in(m_cur, source)) return false;
    if (!_has_prev_for(kind)) return true;
    return !_source_down_in(m_prev, source);
}

bool GameInputReading::was_source_released(int source) const {
    uint32_t kind = _source_snap_kind(source);
    if (!(m_cur.kinds & kind) || !_has_prev_for(kind)) return false;
    return _source_down_in(m_prev, source) && !_source_down_in(m_cur, source);
}

float GameInputReading::get_source_value(int source) const {
    uint32_t kind = _source_snap_kind(source);
    if (!(m_cur.kinds & kind)) return 0.0f;
    int axis = source_bit_index(source, GameInputDevice::SRC_AXIS_LEFT_X, 15);
    if (axis >= 0) return _axis_in(m_cur, axis);
    return _source_down_in(m_cur, source) ? 1.0f : 0.0f;
}

// --- Keyboard ----------------------------------------------------------------

int GameInputReading::get_key_count() const {
    return (m_cur.kinds & gi::SNAP_KEYBOARD) ? (int)m_cur.key_count : 0;
}

int GameInputReading::_find_key(const gi::Snapshot &s, int64_t physical_key) const {
    if (physical_key == 0 || !(s.kinds & gi::SNAP_KEYBOARD)) return -1;
    for (uint32_t i = 0; i < s.key_count && i < gi::kMaxKeys; ++i) {
        if (gi::scan_code_to_key(s.keys[i].scan_code) == physical_key) return (int)i;
    }
    return -1;
}

PackedInt64Array GameInputReading::get_pressed_physical_keys() const {
    PackedInt64Array out;
    if (!(m_cur.kinds & gi::SNAP_KEYBOARD)) return out;
    for (uint32_t i = 0; i < m_cur.key_count && i < gi::kMaxKeys; ++i) {
        int64_t key = gi::scan_code_to_key(m_cur.keys[i].scan_code);
        if (key != 0 && !out.has(key)) out.push_back(key);
    }
    return out;
}

bool GameInputReading::is_physical_key_down(int64_t physical_key) const {
    return _find_key(m_cur, physical_key) >= 0;
}

bool GameInputReading::was_physical_key_pressed(int64_t physical_key) const {
    if (_find_key(m_cur, physical_key) < 0) return false;
    if (!_has_prev_for(gi::SNAP_KEYBOARD)) return true;
    return _find_key(m_prev, physical_key) < 0;
}

bool GameInputReading::was_physical_key_released(int64_t physical_key) const {
    if (!(m_cur.kinds & gi::SNAP_KEYBOARD) || !_has_prev_for(gi::SNAP_KEYBOARD)) return false;
    return _find_key(m_prev, physical_key) >= 0 && _find_key(m_cur, physical_key) < 0;
}

Array GameInputReading::get_key_states() const {
    Array out;
    if (!(m_cur.kinds & gi::SNAP_KEYBOARD)) return out;
    for (uint32_t i = 0; i < m_cur.key_count && i < gi::kMaxKeys; ++i) {
        const gi::KeyData &k = m_cur.keys[i];
        Dictionary d;
        d["scan_code"] = (int64_t)k.scan_code;
        d["virtual_key"] = (int64_t)k.virtual_key;
        d["code_point"] = (int64_t)k.code_point;
        d["is_dead_key"] = k.is_dead_key;
        d["physical_keycode"] = gi::scan_code_to_key(k.scan_code);
        d["keycode"] = gi::virtual_key_to_key(k.virtual_key);
        d["location"] = gi::scan_code_to_location(k.scan_code);
        out.push_back(d);
    }
    return out;
}

// --- Mouse -------------------------------------------------------------------

int GameInputReading::get_mouse_buttons() const {
    return (int)m_cur.mouse.buttons;
}

bool GameInputReading::is_mouse_button_down(int button) const {
    return gi::chord_down(m_cur.mouse.buttons, (uint32_t)button, kMouseButtonMask);
}

bool GameInputReading::was_mouse_button_pressed(int button) const {
    return gi::chord_pressed(m_cur.mouse.buttons, m_prev.mouse.buttons,
                             _has_prev_for(gi::SNAP_MOUSE), (uint32_t)button, kMouseButtonMask);
}

bool GameInputReading::was_mouse_button_released(int button) const {
    return gi::chord_released(m_cur.mouse.buttons, m_prev.mouse.buttons,
                              _has_prev_for(gi::SNAP_MOUSE), (uint32_t)button, kMouseButtonMask);
}

Vector2 GameInputReading::get_mouse_delta() const {
    if (!(m_cur.kinds & gi::SNAP_MOUSE) || !_has_prev_for(gi::SNAP_MOUSE)) return Vector2();
    return Vector2((real_t)gi::wrapping_delta(m_cur.mouse.x, m_prev.mouse.x),
                   (real_t)gi::wrapping_delta(m_cur.mouse.y, m_prev.mouse.y));
}

Vector2 GameInputReading::get_mouse_wheel_delta() const {
    if (!(m_cur.kinds & gi::SNAP_MOUSE) || !_has_prev_for(gi::SNAP_MOUSE)) return Vector2();
    return Vector2((real_t)gi::wrapping_delta(m_cur.mouse.wheel_x, m_prev.mouse.wheel_x),
                   (real_t)gi::wrapping_delta(m_cur.mouse.wheel_y, m_prev.mouse.wheel_y));
}

bool GameInputReading::has_mouse_absolute_position() const {
    return (m_cur.kinds & gi::SNAP_MOUSE) && (m_cur.mouse.positions & 0x1u);
}

Vector2 GameInputReading::get_mouse_absolute_position() const {
    if (!has_mouse_absolute_position()) return Vector2();
    return Vector2((real_t)m_cur.mouse.absolute_x, (real_t)m_cur.mouse.absolute_y);
}

Dictionary GameInputReading::get_mouse_state() const {
    Dictionary d;
    if (!(m_cur.kinds & gi::SNAP_MOUSE)) return d;
    d["buttons"] = (int64_t)m_cur.mouse.buttons;
    d["positions"] = (int64_t)m_cur.mouse.positions;
    d["x"] = m_cur.mouse.x;
    d["y"] = m_cur.mouse.y;
    d["absolute_x"] = m_cur.mouse.absolute_x;
    d["absolute_y"] = m_cur.mouse.absolute_y;
    d["wheel_x"] = m_cur.mouse.wheel_x;
    d["wheel_y"] = m_cur.mouse.wheel_y;
    return d;
}

// --- Sensors -----------------------------------------------------------------

int GameInputReading::get_sensor_kinds() const {
    return (m_cur.kinds & gi::SNAP_SENSORS) ? (int)m_sensor_kinds : 0;
}

Vector3 GameInputReading::get_accelerometer() const {
    if (!(m_cur.kinds & gi::SNAP_SENSORS)) return Vector3();
    const float *a = m_cur.sensors.acceleration_g;
    return Vector3(a[0] * kStandardGravity, a[1] * kStandardGravity, a[2] * kStandardGravity);
}

Vector3 GameInputReading::get_gyroscope() const {
    if (!(m_cur.kinds & gi::SNAP_SENSORS)) return Vector3();
    const float *w = m_cur.sensors.angular_velocity_rad;
    return Vector3(w[0], w[1], w[2]);
}

float GameInputReading::get_heading_degrees() const {
    return (m_cur.kinds & gi::SNAP_SENSORS) ? m_cur.sensors.heading_degrees : 0.0f;
}

int GameInputReading::get_heading_accuracy() const {
    return (m_cur.kinds & gi::SNAP_SENSORS) ? (int)m_cur.sensors.heading_accuracy : 0;
}

Quaternion GameInputReading::get_orientation() const {
    if (!(m_cur.kinds & gi::SNAP_SENSORS)) return Quaternion();
    const float *q = m_cur.sensors.orientation_wxyz;
    return Quaternion(q[1], q[2], q[3], q[0]);
}

// --- Arcade stick, flight stick, racing wheel ----------------------------------

int GameInputReading::get_arcade_stick_buttons() const {
    return (int)(m_cur.arcade_stick.buttons & kSpecialtyButtonMask);
}

int GameInputReading::get_flight_stick_buttons() const {
    return (int)(m_cur.flight_stick.buttons & kSpecialtyButtonMask);
}

int GameInputReading::get_flight_stick_hat() const {
    return (int)m_cur.flight_stick.hat;
}

Vector2 GameInputReading::get_flight_stick_hat_vector() const {
    return GameInputDevice::switch_position_to_vector(m_cur.flight_stick.hat);
}

int GameInputReading::get_racing_wheel_buttons() const {
    return (int)(m_cur.racing_wheel.buttons & kSpecialtyButtonMask);
}

int GameInputReading::get_racing_wheel_gear() const {
    return (int)m_cur.racing_wheel.gear;
}

// --- Raw controller --------------------------------------------------------------

PackedFloat32Array GameInputReading::get_controller_axes() const {
    PackedFloat32Array out;
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_AXIS)) return out;
    for (uint32_t i = 0; i < m_cur.axis_count && i < gi::kMaxControllerAxes; ++i) {
        out.push_back(m_cur.axes[i]);
    }
    return out;
}

Array GameInputReading::get_controller_buttons() const {
    Array out;
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_BUTTON)) return out;
    for (uint32_t i = 0; i < m_cur.button_count && i < gi::kMaxControllerButtons; ++i) {
        out.push_back(gi::controller_button(m_cur, i));
    }
    return out;
}

PackedInt32Array GameInputReading::get_controller_switches() const {
    PackedInt32Array out;
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_SWITCH)) return out;
    for (uint32_t i = 0; i < m_cur.switch_count && i < gi::kMaxControllerSwitches; ++i) {
        out.push_back(m_cur.switches[i]);
    }
    return out;
}

float GameInputReading::get_controller_axis(int index) const {
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_AXIS) || index < 0 ||
            (uint32_t)index >= m_cur.axis_count || (uint32_t)index >= gi::kMaxControllerAxes) {
        return 0.0f;
    }
    return m_cur.axes[index];
}

bool GameInputReading::is_controller_button_down(int index) const {
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_BUTTON) || index < 0) return false;
    return gi::controller_button(m_cur, (uint32_t)index);
}

bool GameInputReading::was_controller_button_pressed(int index) const {
    if (!is_controller_button_down(index)) return false;
    if (!_has_prev_for(gi::SNAP_CONTROLLER_BUTTON)) return true;
    return !gi::controller_button(m_prev, (uint32_t)index);
}

bool GameInputReading::was_controller_button_released(int index) const {
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_BUTTON) || index < 0 ||
            !_has_prev_for(gi::SNAP_CONTROLLER_BUTTON)) {
        return false;
    }
    return gi::controller_button(m_prev, (uint32_t)index) &&
           !gi::controller_button(m_cur, (uint32_t)index);
}

int GameInputReading::get_controller_switch(int index) const {
    if (!(m_cur.kinds & gi::SNAP_CONTROLLER_SWITCH) || index < 0 ||
            (uint32_t)index >= m_cur.switch_count || (uint32_t)index >= gi::kMaxControllerSwitches) {
        return 0;
    }
    return m_cur.switches[index];
}

} // namespace godot
